warehouse/mssql/mssql.go

```diff
@@ -0,0 +1,662 @@
+		primaryKey = column
+	}
+	partitionKey := "id"
+	if column, ok := partitionKeyMap[tableName]; ok {
+		partitionKey = column
+	}
+	var additionalJoinClause string
+	if tableName == warehouseutils.DiscardsTable {
+		additionalJoinClause = fmt.Sprintf(`AND _source.%[3]s = "%[1]s"."%[2]s"."%[3]s" AND _source.%[4]s = "%[1]s"."%[2]s"."%[4]s"`, ms.Namespace, tableName, "table_name", "column_name")
+	}
+	sqlStatement = fmt.Sprintf(`DELETE FROM "%[1]s"."%[2]s" FROM "%[1]s"."%[3]s" as  _source where (_source.%[4]s = "%[1]s"."%[2]s"."%[4]s" %[5]s)`, ms.Namespace, tableName, stagingTableName, primaryKey, additionalJoinClause)
+	pkgLogger.Infof("MS: Deduplicate records for table:%s using staging table: %s\n", tableName, sqlStatement)
+	_, err = txn.Exec(sqlStatement)
+	if err != nil {
+		pkgLogger.Errorf("MS: Error deleting from original table for dedup: %v\n", err)
+		txn.Rollback()
+		return
+	}
+	sqlStatement = fmt.Sprintf(`INSERT INTO "%[1]s"."%[2]s" (%[3]s) SELECT %[3]s FROM ( SELECT *, row_number() OVER (PARTITION BY %[5]s ORDER BY received_at DESC) AS _rudder_staging_row_number FROM "%[1]s"."%[4]s" ) AS _ where _rudder_staging_row_number = 1`, ms.Namespace, tableName, sortedColumnString, stagingTableName, partitionKey)
+	pkgLogger.Infof("MS: Inserting records for table:%s using staging table: %s\n", tableName, sqlStatement)
+	_, err = txn.Exec(sqlStatement)
+
+	if err != nil {
+		pkgLogger.Errorf("MS: Error inserting into original table: %v\n", err)
+		txn.Rollback()
+		return
+	}
+
+	if err = txn.Commit(); err != nil {
+		pkgLogger.Errorf("MS: Error while committing transaction as there was error while loading staging table:%s: %v", stagingTableName, err)
+		return
+	}
+
+	pkgLogger.Infof("MS: Complete load for table:%s", tableName)
+	return
+}
+
+func (ms *HandleT) loadUserTables() (errorMap map[string]error) {
+	errorMap = map[string]error{warehouseutils.IdentifiesTable: nil}
+	pkgLogger.Infof("MS: Starting load for identifies and users tables\n")
+	identifyStagingTable, err := ms.loadTable(warehouseutils.IdentifiesTable, ms.Uploader.GetTableSchemaInUpload(warehouseutils.IdentifiesTable), true)
+	if err != nil {
+		errorMap[warehouseutils.IdentifiesTable] = err
+		return
+	}
+
+	if len(ms.Uploader.GetTableSchemaInUpload(warehouseutils.UsersTable)) == 0 {
+		return
+	}
+	errorMap[warehouseutils.UsersTable] = nil
+
+	unionStagingTableName := misc.TruncateStr(fmt.Sprintf(`%s%s_%s`, stagingTablePrefix, strings.Replace(uuid.NewV4().String(), "-", "", -1), "users_identifies_union"), 127)
+	stagingTableName := misc.TruncateStr(fmt.Sprintf(`%s%s_%s`, stagingTablePrefix, strings.Replace(uuid.NewV4().String(), "-", "", -1), warehouseutils.UsersTable), 127)
+
+	userColMap := ms.Uploader.GetTableSchemaInWarehouse(warehouseutils.UsersTable)
+	var userColNames, firstValProps []string
+	for colName := range userColMap {
+		if colName == "id" {
+			continue
+		}
+		userColNames = append(userColNames, colName)
+		caseSubQuery := fmt.Sprintf(`case
```

Comment:
can you check if MS SQL supports `first_value ignore null` as this query is not performing well for people with postgres destination

Reply:
support not available for mssql server
https://docs.microsoft.com/en-us/sql/t-sql/functions/first-value-transact-sql?view=sql-server-ver15
Only supported in SQL edge. Added same comments in code